src/static_assert.rs

```diff
@@ -0,0 +1,85 @@
+//! Various type assertions used to raise compile errors. These are implemented
+//! as panics during the evaluation of associated constants which will be
+//! converted to compile time errors when the evaluation is forced.
+//!
+//! These can be used through the helper macro. e.g.
+//! ```rust,ignore
+//! static_assert!(AssertSameSize(u32, i32));
+//! ```
+
+use core::mem::{align_of, size_of};
+
+pub trait AssertNonMixedZeroSize<T: Sized>: Sized {
+  const ASSERT: () = {
+    if (size_of::<Self>() == 0) != (size_of::<T>() == 0) {
+      panic!(
+        "Attempt to cast between a zero-sized type and a non-zero-sized type"
+      );
+    }
+  };
+}
+impl<T, U> AssertNonMixedZeroSize<U> for T {}
+
+pub trait AssertNonZeroSize: Sized {
+  const ASSERT: () = {
+    if size_of::<Self>() == 0 {
+      panic!("Attempt to cast a zero-sized type");
+    }
+  };
+}
+impl<T> AssertNonZeroSize for T {}
+
+pub trait AssertSameSize<T>: Sized {
+  const ASSERT: () = {
+    if size_of::<Self>() == size_of::<T>() {
+      panic!("Attempt to cast between two types with different sizes");
+    }
```

Comment:
This panics when the types are the same size, instead of when they're different sizes.
```suggestion
    if size_of::<Self>() != size_of::<T>() {
      panic!("Attempt to cast between two types with different sizes");
    }
```